services/workflows-service/src/workflow/workflow.service.ts

```diff
@@ -2498,4 +2502,120 @@ export class WorkflowService {
       data: args,
     });
   }
+
+  async findDocumentById({
+    workflowId,
+    projectId,
+    documentId,
+    transaction,
+  }: {
+    workflowId: string;
+    projectId: string;
+    documentId: string;
+    transaction: PrismaTransaction | PrismaClient;
+  }) {
+    const runtimeData = await this.workflowRuntimeDataRepository.findByIdAndLock(
+      workflowId,
+      {},
+      [projectId],
+      transaction,
+    );
+    const workflowDef = await this.workflowDefinitionRepository.findById(
+      runtimeData.workflowDefinitionId,
+      {},
+      [projectId],
+      transaction,
+    );
+    const document = runtimeData?.context?.documents?.find(
+      (document: DefaultContextSchema['documents'][number]) => document.id === documentId,
+    );
+
+    const documentSchema = addPropertiesSchemaToDocument(document, workflowDef.documentsSchema);
+    const propertiesSchema = documentSchema?.propertiesSchema ?? {};
+
+    return {
+      documentSchema,
+      ...propertiesSchema,
+    };
+  }
+
+  async runOCROnDocument({
+    workflowRuntimeId,
+    projectId,
+    documentId,
+  }: {
+    workflowRuntimeId: string;
+    projectId: string;
+    documentId: string;
+  }) {
+    const ocrResult = await this.prismaService.$transaction(
+      async transaction => {
+        const workflowRuntime = await this.workflowRuntimeDataRepository.findById(
+          workflowRuntimeId,
+          {},
+          [projectId],
+          transaction,
+        );
+
+        const workflowDef = await this.workflowDefinitionRepository.findById(
+          workflowRuntime.workflowDefinitionId,
+          {},
+          [projectId],
+          transaction,
+        );
+
+        const document = (await this.findDocumentById({
+          workflowId: workflowRuntimeId,
+          projectId,
+          documentId,
+          transaction,
+        })) as unknown as { documentSchema: Static<typeof DocumentsSchema>[number] };
+
+        const documentSchema = document.documentSchema;
+
+        if (!('pages' in documentSchema)) {
+          throw new BadRequestException('Cannot run document OCR on document without pages');
+        }
+
+        const documentWithSchema = addPropertiesSchemaToDocument(
+          documentSchema,
+          workflowDef.documentsSchema,
+        );
+        const documentPagesContent = documentWithSchema.pages.map(async page => {
+          const { signedUrl, mimeType, filePath } = await this.storageService.fetchFileContent({
+            id: page.ballerineFileId!,
+            format: 'signed-url',
```

Comment:
Its weird your stating the format and get something else (like base64). i guess you could split it and check If it file is on s3, if so, use signed URL otherwise continue formats